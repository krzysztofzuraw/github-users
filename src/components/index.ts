export { MainCard } from "@src/components/main-card";
export { List } from "@src/components/list";
export { LocaleCard } from "@src/components/locale-card";
