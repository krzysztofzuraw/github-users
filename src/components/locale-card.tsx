import { Button, Card } from "antd";
import * as React from "react";
import { FormattedMessage } from "react-intl";
import { style } from "typestyle/lib";

interface IProps {
  onClick: (lang: string) => void;
  selectedLang: string;
}

interface IState {
  currentLang: string;
}

const styles = {
  card: style({
    marginTop: "20px",
  }),
  wrapper: style({
    display: "flex",
    justifyContent: "space-between",
  }),
};

export class LocaleCard extends React.Component<IProps, IState> {
  state = { currentLang: this.props.selectedLang };

  handleClick = (lang: string) => {
    this.setState({ currentLang: lang });
    this.props.onClick(lang);
  };

  getButtonType = (lang: string) => {
    const { currentLang } = this.state;
    return currentLang === lang ? "primary" : "default";
  };

  render() {
    const { handleClick, getButtonType } = this;
    return (
      <Card
        title={<FormattedMessage id="locale.card.title" defaultMessage="Change locale" />}
        className={styles.card}
      >
        <div className={styles.wrapper}>
          <Button shape="circle" type={getButtonType("en")} onClick={() => handleClick("en")}>
            EN
          </Button>
          <Button shape="circle" type={getButtonType("pl")} onClick={() => handleClick("pl")}>
            PL
          </Button>
        </div>
      </Card>
    );
  }
}
