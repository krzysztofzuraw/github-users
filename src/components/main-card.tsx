import { Avatar, Button, Card, List } from "antd";
import * as React from "react";
import { FormattedMessage } from "react-intl";

export const MainCard: React.SFC = () => {
  return (
    <Card title={<FormattedMessage id="main.card.header" defaultMessage="Who to follow" />}>
      <List
        itemLayout="horizontal"
        dataSource={["1", "2", "3"]}
        loadMore={
          <div>
            <Button>Refresh</Button>
          </div>
        }
        renderItem={(item: any) => (
          <List.Item actions={[<a>follow</a>]}>
            <List.Item.Meta
              avatar={
                <Avatar src="https://zos.alipayobjects.com/rmsportal/ODTLcjxAfvqbxHnVXCYX.png" />
              }
              title="User Name"
              description="User description goes here"
            />
          </List.Item>
        )}
      />
    </Card>
  );
};
