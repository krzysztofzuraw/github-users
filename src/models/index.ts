export * from "@src/models/user";
export * from "@src/models/locales";
