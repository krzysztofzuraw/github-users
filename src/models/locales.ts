export interface ILocaleDTO {
  language: string;
}

export interface ILocale {
  constructor: {
    create(locale: ILocaleDTO): ILocale;
  };
  language: string;
}

export class Locale implements ILocale {
  // prettier-ignore
  "constructor": typeof Locale;

  constructor(public language: string) {}

  static create(dto: ILocaleDTO): ILocale {
    const model = new Locale(dto.language);
    return model;
  }

  serialize(): ILocaleDTO {
    return {
      language: this.language,
    };
  }
}
