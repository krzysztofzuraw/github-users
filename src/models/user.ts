import cuid from "cuid";

export interface IUserDTO {
  id: string;
  login: string;
  url: string;
  avatar_url: string;
  blog: string;
  bio: string;
  public_repos: number;
  name: string;
}

export interface IUser {
  constructor: {
    create(user: IUserDTO): IUser;
  };

  id: string;
  login: string;
  url: string;
  avatarUrl: string;
  name: string;
  blog: string;
  bio: string;
  publicRepos: number;

  serialize(): IUserDTO;
}

export class User implements IUser {
  // prettier-ignore
  "constructor": typeof User;
  id: string = cuid();

  constructor(
    public login: string,
    public url: string,
    public avatarUrl: string,
    public name: string,
    public blog: string,
    public publicRepos: number,
    public bio: string
  ) {}

  static create(dto: IUserDTO): IUser {
    const model = new User(
      dto.login,
      dto.url,
      dto.avatar_url,
      dto.name,
      dto.blog,
      dto.public_repos,
      dto.bio
    );
    model.id = dto.id;
    return model;
  }

  serialize(): IUserDTO {
    return {
      avatar_url: this.avatarUrl,
      bio: this.bio,
      blog: this.blog,
      id: this.id,
      login: this.login,
      name: this.name,
      public_repos: this.publicRepos,
      url: this.url,
    };
  }
}
