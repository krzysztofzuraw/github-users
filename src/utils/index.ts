export { default as componentWithIntl } from "@src/utils/tests";
