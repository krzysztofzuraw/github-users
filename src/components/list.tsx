import { Avatar, Button, List as AntList } from "antd";
import * as React from "react";
import { FormattedMessage } from "react-intl";
import { style as tss } from "typestyle/lib";

import { IUser } from "@src/models";

interface IProps {
  items: IUser[];
  loading: boolean;
  onReload: () => void;
}

const styles = {
  actions: tss({
    alignItems: "center",
    display: "flex",
    justifyContent: "center",
    marginTop: "10px",
  }),
};

export const List: React.SFC<IProps> = ({ items, loading, onReload }) => (
  <AntList
    itemLayout="horizontal"
    dataSource={items}
    loadMore={
      <div className={styles.actions}>
        <Button onClick={onReload} loading={loading}>
          <FormattedMessage id="list.button.reload" defaultMessage="Refresh" />
        </Button>
      </div>
    }
    loading={loading}
    renderItem={(item: IUser) => (
      <AntList.Item
        actions={[
          <a href={item.htmlUrl}>
            <FormattedMessage id="list.link.profile" defaultMessage="Profile" />
          </a>,
        ]}
      >
        <AntList.Item.Meta
          avatar={<Avatar src={item.avatarUrl} />}
          title={item.login}
          description={
            item.bio || (
              <FormattedMessage id="list.description.default" defaultMessage="No description" />
            )
          }
        />
      </AntList.Item>
    )}
  />
);
