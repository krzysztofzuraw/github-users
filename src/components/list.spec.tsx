import * as React from "react";
import { cleanup, render } from "react-testing-library";

import { List } from "@src/components";
import { User } from "@src/models";
import { componentWithIntl } from "@src/utils";

afterEach(cleanup);

describe("List component", () => {
  it("should render empty list when there is no items to load", () => {
    const component = render(
      componentWithIntl(<List items={[]} loading={false} onReload={jest.fn()} />)
    );
    expect(component).toMatchSnapshot();
  });

  it("should render items list", () => {
    const users = User.create({
      avatar_url: "https://",
      bio: "bio",
      blog: "blog",
      html_url: "https://",
      id: "1",
      public_repos: 1,
      name: "name",
      login: "user",
    });
    const component = render(
      componentWithIntl(<List items={[users]} loading={false} onReload={jest.fn()} />)
    );
    expect(component).toMatchSnapshot();
  });
});
