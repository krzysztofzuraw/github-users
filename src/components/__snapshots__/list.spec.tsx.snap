// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List component should render empty list when there is no items to load 1`] = `
Object {
  "baseElement": <body>
    <div>
      <div
        class="ant-list ant-list-split ant-list-something-after-last-item"
      >
        <div
          class="ant-spin-nested-loading"
        >
          <div
            class="ant-spin-container"
          >
            <div
              class="ant-list-empty-text"
            >
              No data
            </div>
          </div>
        </div>
        <div
          class="f9r8yry"
        >
          <button
            class="ant-btn"
            type="button"
          >
            <span>
              Refresh
            </span>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ant-list ant-list-split ant-list-something-after-last-item"
    >
      <div
        class="ant-spin-nested-loading"
      >
        <div
          class="ant-spin-container"
        >
          <div
            class="ant-list-empty-text"
          >
            No data
          </div>
        </div>
      </div>
      <div
        class="f9r8yry"
      >
        <button
          class="ant-btn"
          type="button"
        >
          <span>
            Refresh
          </span>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;

exports[`List component should render items list 1`] = `
Object {
  "baseElement": <body>
    <div>
      <div
        class="ant-list ant-list-split ant-list-something-after-last-item"
      >
        <div
          class="ant-spin-nested-loading"
        >
          <div
            class="ant-spin-container"
          >
            <div
              class="ant-list-item"
            >
              <div
                class="ant-list-item-meta"
              >
                <div
                  class="ant-list-item-meta-avatar"
                >
                  <span
                    class="ant-avatar ant-avatar-circle ant-avatar-image"
                  >
                    <img
                      src="https://"
                    />
                  </span>
                </div>
                <div
                  class="ant-list-item-meta-content"
                >
                  <h4
                    class="ant-list-item-meta-title"
                  >
                    user
                  </h4>
                  <div
                    class="ant-list-item-meta-description"
                  >
                    bio
                  </div>
                </div>
              </div>
              <ul
                class="ant-list-item-action"
              >
                <li>
                  <a
                    href="https://"
                  >
                    <span>
                      Profile
                    </span>
                  </a>
                </li>
              </ul>
            </div>
          </div>
        </div>
        <div
          class="f9r8yry"
        >
          <button
            class="ant-btn"
            type="button"
          >
            <span>
              Refresh
            </span>
          </button>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="ant-list ant-list-split ant-list-something-after-last-item"
    >
      <div
        class="ant-spin-nested-loading"
      >
        <div
          class="ant-spin-container"
        >
          <div
            class="ant-list-item"
          >
            <div
              class="ant-list-item-meta"
            >
              <div
                class="ant-list-item-meta-avatar"
              >
                <span
                  class="ant-avatar ant-avatar-circle ant-avatar-image"
                >
                  <img
                    src="https://"
                  />
                </span>
              </div>
              <div
                class="ant-list-item-meta-content"
              >
                <h4
                  class="ant-list-item-meta-title"
                >
                  user
                </h4>
                <div
                  class="ant-list-item-meta-description"
                >
                  bio
                </div>
              </div>
            </div>
            <ul
              class="ant-list-item-action"
            >
              <li>
                <a
                  href="https://"
                >
                  <span>
                    Profile
                  </span>
                </a>
              </li>
            </ul>
          </div>
        </div>
      </div>
      <div
        class="f9r8yry"
      >
        <button
          class="ant-btn"
          type="button"
        >
          <span>
            Refresh
          </span>
        </button>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "getAllByAltText": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getAllByValue": [Function],
  "getByAltText": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "getByValue": [Function],
  "queryAllByAltText": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryAllByValue": [Function],
  "queryByAltText": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "queryByValue": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
